packages/deployments/contracts/contracts/shared/libraries/Unwrapper.sol

```diff
@@ -0,0 +1,159 @@
+
+  /**
+   * @notice The wrapper contract that this contract will always use for unwrapping native token.
+   */
+  Wrapper public immutable WRAPPER;
+
+  // ============= Modifiers ==============
+
+  /**
+   * @notice Ensure caller is only the designated CONNEXT bridge address.
+   */
+  modifier onlyConnext() {
+    require(msg.sender == CONNEXT, "unwrap: !connext");
+    _;
+  }
+
+  // ============ Constructor ============
+
+  constructor(address connext, address wrapper) {
+    WRAPPER = Wrapper(wrapper);
+    CONNEXT = connext;
+  }
+
+  // ================ Getters ================
+
+  /**
+   * @notice Read method to get the target wrapper contract. Make sure this is the wrapper
+   * you're looking for!
+   */
+  function getTargetWrapperContract() external view returns (address) {
+    return address(WRAPPER);
+  }
+
+  /**
+   * @notice Read method to get Connext bridge address, the only address permissioned to call
+   * `xReceive` below.
+   */
+  function getConnext() external view returns (address) {
+    return address(CONNEXT);
+  }
+
+  // ============ Public Functions ============
+
+  /**
+   * @notice xReceive implementation for receiving cross-chain calls from Connext.
+   * @dev We mostly ignore `originSender` argument: this could be a contract or EOA, but our
+   * recipient should be specified in our `callData`! We only fallback to using `originSender` IFF
+   * recipient argument is missing.
+   * @dev If unwrapping (i.e. `withdraw`) fails, will emit UnwrappingFailed event! We will attempt
+   * to transfer the wrapped tokens to the
+   *
+   * @param amount - The amount to transfer. Should NOT be 0, or this call will revert.
+   * @param asset - This *should be* the wrapper contract address, an ERC20 token approved by the
+   * Connext bridge. IFF this does NOT match the WRAPPER contract address stored in this contract,
+   * we'll try to `IERC20.transfer` the assets to the intended recipient. If even that fails, we'll
+   * orphan the received tokens here for later pickup.
+   * @param originSender - Not used except in the edge case where the `callData` is missing a valid
+   * intended recipient address.
+   * @param callData - Should be a tuple of just `(address)`. The address is the intended
+   * recipient of the unwrapped native tokens. Whether it's ether or wether (i.e. whether it's
+   * wrapped native tokens or native tokens) depends on whether we succeeded in the unwrapping
+   * process.
+   */
+  function xReceive(
+    bytes32,
+    uint256 amount,
+    address asset,
+    address originSender,
+    uint32,
+    bytes memory callData
+  ) external onlyConnext returns (bytes memory) {
+    // Sanity check: amount is non-zero (otherwise, what are we unwrapping?).
+    require(amount != 0, "unwrap: !amount");
+
+    // Get the target recipient, which should be in the callData.
+    // TODO: What if decode results in missing address? i.e. empty callData or poorly formatted callData?
```

Comment:
This is fine, we can't fully control for this. However it does become problematic because we will try to execute it forever. This is where the exponential backoff comes into play.